src/remediations/fifi_2.js

```diff
@@ -217,3 +219,112 @@ async function storePlaybookRun (playbook_run_id, remediation, username) {
 
     await queries.insertRHCPlaybookRun(run);
 }
+
+/**
+ * Sync dispatcher_runs table with playbook-dispatcher API for specific playbook runs
+ * 
+ * Logic:
+ * - If any dispatcher_run is failed or timeout, skip sync (no API call needed)
+ * - If pending/running dispatcher_runs exist OR no dispatcher_runs exist, sync with API
+ * - For existing runs: update status, for missing runs: create new entries
+ */
+exports.syncDispatcherRunsForPlaybookRuns = async function (playbookRunIds) {
+    if (!playbookRunIds.length) {
+        return [];
+    }
+    
+    try {
+        const sql = `
+            SELECT 
+                pr.id as playbook_run_id,
+                COUNT(dr.dispatcher_run_id)::int as total_dispatcher_runs,
+                COUNT(CASE WHEN dr.status IN ('failure', 'timeout') THEN 1 END)::int as failed_runs,
+                COUNT(CASE WHEN dr.status IN ('pending', 'running') THEN 1 END)::int as incomplete_runs
+            FROM playbook_runs pr
+            LEFT JOIN dispatcher_runs dr ON pr.id = dr.remediations_run_id
+            WHERE pr.id IN (:playbookRunIds)
+            GROUP BY pr.id
+        `;
+
+        const playbookRunCounts = await db.s.query(sql, {
+            replacements: { playbookRunIds },
+            type: db.s.QueryTypes.SELECT
+        });
+
+        const needsStatusSync = [];
+        const needsBackfill = [];
+
+        for (const runCounts of playbookRunCounts) {
+            if (runCounts.total_dispatcher_runs === 0) {
+                // No dispatcher_runs exist, needs backfill
+                needsBackfill.push(runCounts.playbook_run_id);
+            } else if (runCounts.incomplete_runs > 0 && runCounts.failed_runs === 0) {
+                // Has pending/running, needs update
+                needsStatusSync.push(runCounts.playbook_run_id);
+            }
+        }
+
+        const runsToSync = [...needsStatusSync, ...needsBackfill];
+
+        if (runsToSync.length > 0) {
+            // Limit concurrent API calls
+            const limit = pLimit(10);
+            
+            await Promise.all(runsToSync.map(playbookRunId => limit(async () => {
+                try {
+                    const filter = {
+                        filter: { 
+                            service: 'remediations',
+                            labels: { 'playbook-run': playbookRunId } 
+                        }
+                    };
+                    const fields = {
+                        fields: { data: ['id', 'status'] }
+                    };
```

Comment:
Ooh... the Sorcery comment above is actually pretty good.  Let me know if it's confusing (it's basically a way to return an array that only contains the runs that were successfully synced).

Reply:
Yep I like that too I made that change 🙂